react/features/face-centering/middleware.js

```diff
@@ -0,0 +1,135 @@
+    const isInTileView = shouldDisplayTileView(state);
+
+    if (!getCurrentConference(getState())) {
+        return next(action);
+    }
+
+    if (action.type === CONFERENCE_JOINED) {
+        if (faceCoordinatesSharing?.enabled) {
+            dispatch(loadWorker());
+
+            if (isInTileView) {
+                dispatch(startFaceRecognition());
+            }
+        }
+
+        // allow using remote face centering data when local face centering is not enabled
+        action.conference.on(
+            JitsiConferenceEvents.ENDPOINT_MESSAGE_RECEIVED,
+            (...args) => {
+                if (args && args.length >= 2) {
+                    const [ { _id }, eventData ] = args;
+
+                    if (eventData.type === FACE_BOX_EVENT_TYPE) {
+                        dispatch({
+                            type: UPDATE_FACE_COORDINATES,
+                            faceBox: eventData.faceBox,
+                            id: _id
+                        });
+                    }
+                }
+            });
+
+        return next(action);
+    }
+
+    if (!faceCoordinatesSharing?.enabled) {
+        return next(action);
+    }
+
+    // might need to change logic here if object-fit: cover will be used outside of tileview in the future
+    if (action.type === SET_TILE_VIEW) {
+        if (action.enabled) {
+            dispatch(startFaceRecognition());
+        } else {
+            dispatch(stopFaceRecognition());
+        }
+
+        return next(action);
+    }
+
+    if (!isInTileView) {
+        return next(action);
+    }
+
+    switch (action.type) {
+    case CONFERENCE_WILL_LEAVE : {
+        dispatch(stopFaceRecognition());
+
+        return next(action);
+    }
+    case TRACK_UPDATED: {
+        const { videoType, type } = action.track.jitsiTrack;
+
+        if (videoType === 'camera') {
+            const { muted, videoStarted } = action.track;
+
+            if (videoStarted === true) {
+                dispatch(startFaceRecognition());
+            }
+            if (muted !== undefined) {
+                if (muted) {
+                    dispatch(stopFaceRecognition());
+                } else {
+                    dispatch(startFaceRecognition());
+                    type === 'presenter' && changeTrack(action.track);
+                }
+            }
+        }
+
+        return next(action);
+    }
+    case TRACK_ADDED: {
+        const { mediaType, videoType } = action.track;
+
+        if (mediaType === 'presenter' && videoType === 'camera') {
+            dispatch(startFaceRecognition());
+            changeTrack(action.track);
+        }
+
+        return next(action);
+    }
+    case TRACK_REMOVED: {
+        const { videoType } = action.track.jitsiTrack;
+
+        if ([ 'camera', 'desktop' ].includes(videoType)) {
+            dispatch(stopFaceRecognition());
+        }
+
+        return next(action);
+    }
+    case VIRTUAL_BACKGROUND_TRACK_CHANGED: {
```

Comment:
Is this for handling the case when we use the weatherman mode?

Reply:
that's what I assumed, this is taken from the emotions implementation

Reply:
talked to Gabi, we need to use the updated track after a virtual background change

Reply:
removed for now